### PR TITLE
8317050: Scope should reflect lifetime of underying resource

### PR DESCRIPTION
**Pull request by mcimadamore**
This patch addresses some issues with `MemorySegment.Scope::equals`. More specifically, when two segments are created from the same array/buffer, their underlying scope should be equal, but currently it is not, which is confusing.

This patch addesses that by adding a new subclass of `GlobalSession` (`HeapSession`) whose `equals`/`hashcode` methods are defined in terms of the heap reference it wraps. Note that the reference is deduped on construction - that is, we always try to find the "bottom" refrence that we should try to keep alive with the session, and we use that for comparisons (which avoid issues when comparing buffer slices and views).

As part of the changes, this patch also adds more javadoc on `MemorySegment.Scope` to document how equality works. And, this patch also makes all zero-length memory segments associate with the same "native" scope. This is a scope that is always alive, but is not the global scope (not to confuse with segments created by the global arena). While what the implementation does today is technically correct (a new fresh scope is returned with each new ZLMS), it seems also overkill given that the scope associated with segments originated outside Java code is not very interesting - if users really care about lifetime of such segments they should override the scope with `MemorySegment::reinterpret`. This leads to some changes in the javadoc of other methods where I replaced `fresh scope that is always alive` with just `scope that is always alive`, to avoid suggesting that a new scope should be created each time.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed (1 review required, with at least 1 [Committer](https://openjdk.org/bylaws#committer))

### Issue
 * [JDK-8317050](https://bugs.openjdk.org/browse/JDK-8317050): Scope should reflect lifetime of underying resource (**Enhancement** - P2)

### Reviewers
 * [Jorn Vernee](https://openjdk.org/census#jvernee) (@JornVernee - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/panama-foreign.git pull/898/head:pull/898` \
`$ git checkout pull/898`

Update a local copy of the PR: \
`$ git checkout pull/898` \
`$ git pull https://git.openjdk.org/panama-foreign.git pull/898/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 898`

View PR using the GUI difftool: \
`$ git pr show -t 898`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/panama-foreign/pull/898.diff">https://git.openjdk.org/panama-foreign/pull/898.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/panama-foreign/pull/898#issuecomment-1737582076)

### CODE REVIEW

File: test/jdk/java/foreign/TestScope.java [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+import java.lang.foreign.Arena;
+import java.lang.foreign.MemorySegment;
+import java.lang.foreign.SymbolLookup;
+import java.nio.ByteBuffer;
+import java.nio.IntBuffer;
+
+import static org.testng.Assert.*;
+
+public class TestScope {
```
Comment (JornVernee) [inline test/jdk/java/foreign/TestScope.java:39]: We already have `TestMemorySession`. Maybe these two test classes can be merged.

File: src/java.base/share/classes/jdk/internal/foreign/GlobalSession.java [hunk 2]
```diff
@@ -67,4 +70,52 @@ void addInternal(ResourceList.ResourceCleanup resource) {
...
+        private static Object base(Object o) {
+            if (o instanceof DirectBuffer directBuffer) {
+                return directBuffer.attachment() != null ?
+                        directBuffer.attachment() : directBuffer;
+            } else if (o instanceof Buffer heapBuffer) {
+                return NIO_ACCESS.getBufferBase(heapBuffer);
+            } else {
+                return o;
+            }
+        }
```
[the remark this answers is not in the document]
Comment (mcimadamore) [reply src/java.base/share/classes/jdk/internal/foreign/GlobalSession.java:119]: good point

File: test/jdk/java/foreign/TestScope.java (hunk 1 above)
Comment (mcimadamore) [reply test/jdk/java/foreign/TestScope.java:39]: Uhmmm - not sure. TestMemorySession is really about close(). In that respect, it is more of an Arena test than it is about a scope test. This new test is about asserting properties on the scope of some segments.

**Pull request closed without merging**